src/C/sahap-c.c

```diff
@@ -148,6 +152,188 @@ static void InitializeSystem(void) {
     }
 }
 
+
+// NEW CODE STARTS HERE: 
+
+
+// Extract k-mer from the tail of a read (null-terminated)
+char* tailKmer(const READ* r, int k) {
+    if (!r || !r->let || k > r->numSites) return NULL;
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    int startIdx = r->numSites - k;
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[startIdx + i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Extract k-mer from the head of a read (null-terminated)
+char* headKmer(const READ* r, int k) {
+    if (!r || !r->let || k > r->numSites) return NULL;
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Check if two k-mers match (string compare)
+bool kmersMatch(const char* kmer1, const char* kmer2) {
+    if (!kmer1 || !kmer2) return false;
+    return strcmp(kmer1, kmer2) == 0;
+}
+
+// Compare two reads for equality
+bool readsEqual(const READ* a, const READ* b) {
+    if (a->numSites != b->numSites) return false;
+    if (a->firstSite != b->firstSite) return false;
+    for (int i = 0; i < a->numSites; ++i) {
+        if (a->let[i] != b->let[i]) return false;
+    }
+    return true;
+}
+
+// Append one base to the right of a read
+READ addReadsRight(const READ* base, char a) {
+    READ out = *base; 
```

Comment:
OK this is weird to me. Passing in the pointer to a READ, and then deferencing the pointer means you've created a copy of the entire READ into the variable called "out"... but not this will not be a "deep" copy because the letters array is itself a pointer and only the POINTER to the "letters" array will be copied, not the array itself. So you're getting yourself into murky waters here.

It would probably make more sense to use "realloc" on the letters array... but that's also not particularly efficient.